File: src/client/java/net/greenjab/fixedminecraft/map_book/MapBookScreen.java
package net.greenjab.fixedminecraft.map_book;

import net.greenjab.fixedminecraft.network.MapBookPlayer;
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookItem;
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookState;
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookStateManager;
import net.greenjab.fixedminecraft.registry.item.map_book.MapStateData;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.gl.RenderPipelines;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.widget.ButtonWidget;
import net.minecraft.client.network.PlayerListEntry;
import net.minecraft.client.texture.Sprite;
import net.minecraft.component.DataComponentTypes;
import net.minecraft.component.type.MapIdComponent;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.item.map.MapDecoration;
import net.minecraft.item.map.MapDecorationTypes;
import net.minecraft.item.map.MapState;
import net.minecraft.scoreboard.Team;
import net.minecraft.screen.ScreenTexts;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.ColorHelper;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import org.joml.Matrix3x2fStack;

import java.awt.*;
import java.util.ArrayList;
import java.util.ConcurrentModificationException;
import java.util.Iterator;
import java.util.Objects;
import java.util.Optional;

/** Credit: Nettakrim */
public class MapBookScreen extends Screen {
    ItemStack item;
    public float x = 0.0f;
    public float y = 0.0f;
    public float scale = 1.0f;
    private float targetScale = 0.5f;

    public MapBookScreen(ItemStack item){
        super(item.getName());
        this.item = item;
    }

    @Override public void init() {
        if (client != null && client.player != null) {
            x = (float) -client.player.getX();
            y = (float) -client.player.getZ();
        }
        setScale(targetScale, width/2.0f, height/2.0f);

        for (int i = 4;i >= 0;i--) {
            for (MapStateData mapStateData : MapBookItem.getMapStates(item, client.world)) {
                if (mapStateData.mapState.scale == i) addDrawable(new MapTile(this, mapStateData.id, mapStateData.mapState, client));
            }
        }

        addDrawableChild(ButtonWidget.builder(ScreenTexts.DONE, button -> { this.close();})
                .dimensions(width / 2 - 100, height -40, 200, 20).build());
    }

    @Override public boolean shouldPause() {
        return false;
    }

    @Override public boolean mouseClicked(double mouseX, double mouseY, int button, boolean doubleClick) {
        if (button == 0 && Screen.hasShiftDown()) {
            int id = getMapBookId(item);
            if (id != -1) {
                MapBookPlayer marker = MapBookStateManager.INSTANCE.getClientMapBookState(getMapBookId(item)).marker;

                var pos = new Vec3d(mouseX, mouseY, 0.0);
                pos = pos.multiply((1 / scale));
                pos = pos.subtract(width / 2.0, height / 2.0, 0.0);
                pos = pos.subtract(this.x / scale, this.y / scale, 0.0);
                String dim = client.world.getDimensionEntry().getIdAsString();
                if (!marker.dimension.contains(dim) || (pos.distanceTo(new Vec3d(marker.x, marker.z, 0)) * scale)>5) {
                    MinecraftClient.getInstance().getNetworkHandler().sendChatCommand(String.format(
                            "mapBookMarker %s \"%s\" \"%s\" \"%s\"",
                            getMapBookId(item), pos.getX(), pos.getY(), dim));
                } else {
                    MinecraftClient.getInstance().getNetworkHandler().sendChatCommand(String.format(
                            "mapBookMarker %s \"%s\" \"%s\" \"%s\"",
                            getMapBookId(item), 0, 0, ""));
                }
            }
        }
        if (button == 2) {
            if (client.player.getAbilities().creativeMode) {
                var pos = new Vec3d(mouseX, mouseY, 0.0);
                pos = pos.multiply((1/scale));
                pos = pos.subtract(width / 2.0, height / 2.0, 0.0);
                pos = pos.subtract(this.x/scale, this.y/scale, 0.0);
                MinecraftClient.getInstance().getNetworkHandler().sendChatCommand(String.format(
                        "tp %.6f %.6f %.6f",
                        pos.getX(), client.player.getY(), pos.getY()));
            }
        }
        return super.mouseClicked(mouseX, mouseY, button, doubleClick);
    }

    @Override public boolean mouseDragged(double mouseX, double mouseY, int button, double deltaX, double deltaY) {
        if (button < 2 && !Screen.hasShiftDown()) {
            x += deltaX;
            y += deltaY;
        }
        return super.mouseDragged(mouseX, mouseY, button, deltaX, deltaY);
    }

    @Override public boolean mouseScrolled(double mouseX, double mouseY, double horizontalAmount, double verticalAmount) {
        if (verticalAmount != 0.0) {
            targetScale = zoom(scale, (float) -verticalAmount);
        }
        return super.mouseScrolled(mouseX, mouseY, horizontalAmount, verticalAmount);
    }

    @Override public void render(DrawContext context, int mouseX, int mouseY, float delta) {
        if (context == null) return;

        if (scale != targetScale) {
            float newScale = MathHelper.lerp(delta, scale, targetScale);
            setScale(newScale, mouseX, mouseY);
        }

        super.render(context, mouseX, mouseY, delta);

        PlayerEntity thisPlayer = client.player;
        if (thisPlayer==null)return;

        int id = getMapBookId(item);

        if (id != -1) {
            renderIcons(context);
            MapBookPlayer p = new MapBookPlayer();
            p.setPlayer(thisPlayer);

            MapBookPlayer marker = MapBookStateManager.INSTANCE.getClientMapBookState(id).marker;
            if (marker.dimension.contains(p.dimension)) renderMarker(context, marker);

            ArrayList<MapBookPlayer> m = MapBookStateManager.INSTANCE.getClientMapBookState(id).players;
            if (m != null) {
                try {
                    for (MapBookPlayer player : m) {
                        if (player.dimension.contains(p.dimension)) {
                            if (!(player.name.contains(p.name) && p.name.contains(player.name))) {
                                renderPlayerIcon(context, player, false);
                            }
                        }
                    }
                } catch (ConcurrentModificationException ignored) {
                }
            }
            renderPlayerIcon(context, p, true);
            renderPosition(context, mouseX, mouseY);
        }
    }

    private void renderPosition(DrawContext context, int mouseX, int mouseY) {

        var pos = new Vec3d(mouseX, mouseY, 0.0);
        pos = pos.multiply((1/scale));
        pos = pos.subtract(width / 2.0, height / 2.0, 0.0);
        pos = pos.subtract(this.x/scale, this.y/scale, 0.0);


        TextRenderer textRenderer = MinecraftClient.getInstance().textRenderer;
        String text = (int)pos.getX() + ", " + (int)pos.getY();
        int o = textRenderer.getWidth(text);
        Objects.requireNonNull(textRenderer);
        Matrix3x2fStack matrix = context.getMatrices();
        matrix.pushMatrix();
        matrix.translate((int)((width / 2.0f) -o / 2f), (int)(height -60.0f + 8f));
        context.fill(- 1, - 1, o, 9 , (new Color(50, 50, 50, 150)).hashCode());
        context.drawText(textRenderer, text, 0, 0, -1, true);
        matrix.popMatrix();

    }

    private void renderPlayerIcon(DrawContext context, MapBookPlayer player, boolean thisPlayer) {

        MinecraftClient minecraftClient = MinecraftClient.getInstance();
        float x = (float) player.x;
        float z = (float) player.z;
        float rotation = player.yaw;
        Matrix3x2fStack matrix = context.getMatrices();

        matrix.pushMatrix();
        matrix.translate(this.x, this.y);
        matrix.scale(this.scale, this.scale);
        matrix.translate(x + width/ 2.0f, z + height / 2.0f);
        matrix.rotate((float) (rotation * Math.PI/180.0f));
        matrix.rotate((float) (Math.PI));
        matrix.scale(8.0f, 8.0f);
        matrix.scale(2.5f, 2.5f);
        matrix.scale(1f / this.scale, 1f / this.scale);
        matrix.translate(-0.5f, -0.5f);

        if (thisPlayer) {
            context.drawGuiTexture(RenderPipelines.GUI_TEXTURED, Identifier.of(
                            //"hud/locator_bar_dot/map_decorations/" + mapIcon.getAssetId().getPath()),
                            "hud/locator_bar_dot/map_decorations/player"),
                    0, 0, 1, 1, -1);
        } else {
            int color = ColorHelper.withBrightness(ColorHelper.withAlpha(255, player.name.hashCode()), 0.9F);
            for (PlayerListEntry playerListEntry : client.player.networkHandler.getPlayerList()) {
                if (Objects.equals(playerListEntry.getProfile().getName(), player.name)) {
                    Team team = playerListEntry.getScoreboardTeam();
                    if (team != null) {
                        Formatting formatting = team.getColor();
                        if (formatting.isColor()) {
                            color = (new Color(formatting.getColorValue().intValue()).hashCode());
                        }
                    }
                }
            }
            context.drawGuiTexture(RenderPipelines.GUI_TEXTURED, Identifier.of(
                            "hud/locator_bar_dot/map_decorations/player"),
                    0, 0, 1, 1, color);
        }
        matrix.popMatrix();

        TextRenderer textRenderer = minecraftClient.textRenderer;
        String text = player.name;
        int o = textRenderer.getWidth(text);
        matrix.pushMatrix();

        matrix.translate(this.x, this.y);
        matrix.scale(this.scale, this.scale);
        matrix.translate(x + width / 2.0f, z + height / 2.0f);
        matrix.scale(1 / this.scale, 1 / this.scale);
        matrix.translate(-o / 2f, 10.0f);

        context.fill(- 1, - 1, o, 9, (new Color(50, 50, 50, 150)).hashCode());
        context.drawText(textRenderer, text, 0, 0, -1, true);
        matrix.popMatrix();

    }

    private void renderIcons(DrawContext context) {

        for (MapStateData mapStateData : getMapStates(item, client.world)) {
            float render = 0.0f;
            if (client.world.getDimensionEntry().getIdAsString().contains(mapStateData.mapState.dimension.getValue().toString()))
                render = 1.0f;
            if (client.world.getDimensionEntry().getIdAsString().contains("the_nether") && mapStateData.mapState.dimension.getValue().toString().contains("overworld"))
                render = 1/8.0f;
            if (render>0) {
                Iterator<MapDecoration>  var11 = mapStateData.mapState.getDecorations().iterator();
                Matrix3x2fStack matrix = context.getMatrices();
                while (var11.hasNext()) {
                    MapDecoration mapIcon = var11.next();
                    if (!mapIcon.type().getIdAsString().contains("player")) {
                        matrix.pushMatrix();
                        matrix.translate(this.x, this.y);
                        matrix.scale(this.scale, this.scale);
                        float mapScale = (float) Math.pow(2, mapStateData.mapState.scale);
                        float offset = 64f * mapScale;
                        float x = (mapStateData.mapState.centerX - offset + (mapIcon.x() + 128 + 1) * mapScale / 2) * render;
                        float z = (mapStateData.mapState.centerZ - offset + (mapIcon.z() + 128 + 1) * mapScale / 2) * render;
                        matrix.translate(x + width / 2.0f, z + height / 2.0f);
                        matrix.scale(8.0f, 8.0f);
                        matrix.scale(2.5f, 2.5f);
                        matrix.scale(1f / this.scale, 1f / this.scale);
                        matrix.translate(-0.5f, -0.5f);

                        context.drawGuiTexture(RenderPipelines.GUI_TEXTURED, Identifier.of(
                                        "hud/locator_bar_dot/map_decorations/" + mapIcon.getAssetId().getPath()),
                                0, 0, 1, 1, -1);
                        matrix.popMatrix();

                        if (mapIcon.name().isPresent()) {

                            TextRenderer textRenderer = MinecraftClient.getInstance().textRenderer;
                            Text text = mapIcon.name().get();
                            int o = textRenderer.getWidth(text);
                            matrix.pushMatrix();

                            matrix.translate(this.x, this.y);
                            matrix.scale(this.scale, this.scale);
                            float mapx = (mapStateData.mapState.centerX - offset + (mapIcon.x() + 128 + 1) * mapScale / 2) * render;
                            float mapz = (mapStateData.mapState.centerZ - offset + (mapIcon.z() + 128 + 1) * mapScale / 2) * render;
                            matrix.translate(mapx + width / 2.0f, mapz + height / 2.0f);
                            matrix.scale(1 / this.scale, 1 / this.scale);
                            matrix.translate(-o / 2f, 10.0f);

                            context.fill(- 1, - 1, o, 9, (new Color(50, 50, 50, 150)).hashCode());
                            context.drawText(textRenderer, text, 0, 0, -1, true);
                            matrix.popMatrix();
                        }
                    }
                }
            }
        }
    }

    private void renderMarker(DrawContext context, MapBookPlayer player) {
        float x = (float) player.x;
        float z = (float) player.z;
        float rotation = player.yaw;
        Matrix3x2fStack matrix = context.getMatrices();

        matrix.pushMatrix();
        matrix.translate(this.x, this.y);
        matrix.scale(this.scale, this.scale);
        matrix.translate(x + width/ 2.0f, z + height / 2.0f);
        matrix.rotate((float) (rotation * Math.PI/180.0f));
        matrix.rotate((float) (Math.PI));
        matrix.scale(8.0f, 8.0f);
        matrix.scale(2.5f, 2.5f);
        matrix.scale(1f / this.scale, 1f / this.scale);
        matrix.translate(-0.5f, -0.5f);

        context.drawGuiTexture(RenderPipelines.GUI_TEXTURED, Identifier.of(
                        "hud/locator_bar_dot/map_decorations/target_x"),
                0, 0, 1, 1, -1);
        matrix.popMatrix();
    }


    private ArrayList<MapStateData> getMapStates(ItemStack stack, World world) {
        ArrayList<MapStateData> list = new ArrayList<>();
        MapBookState mapBookState = getMapBookState(stack, world);

        if (mapBookState != null) {
            for (int i : mapBookState.mapIDs) {
                MapState mapState = world.getMapState(new MapIdComponent(i));
                if (mapState != null) {
                    list.add(new MapStateData(new MapIdComponent(i), mapState));
                }
            }
        }
        return list;
    }

    private MapBookState getMapBookState(ItemStack stack, World world) {
        int id = getMapBookId(stack) ;
        if (id == -1) return null;
        return MapBookStateManager.INSTANCE.getClientMapBookState(id);
    }

    private static int getMapBookId(ItemStack stack) {
        MapIdComponent mapIdComponent = stack.getOrDefault(DataComponentTypes.MAP_ID, null);
        if (mapIdComponent!=null) return mapIdComponent.id();
        return -1;
    }

    private void setScale(float newScale, float mouseX, float mouseY) {
        float offsetX = x-mouseX;
        float offsetY = y-mouseY;

        float scaleChange = newScale/scale;

        x = (scaleChange * offsetX)+mouseX;
        y = (scaleChange * offsetY)+mouseY;

        scale = newScale;
    }

    private float zoom(float start, float scroll) {
        // logarithmic zoom that doesn't drift when zooming in and out repeatedly
        float absScroll = Math.abs(scroll);
        float speed = 5.0f;
        float newZoom =  scroll > 0 ? start - (start / (scroll * speed)) : (start * absScroll * speed) / (absScroll * speed - 1);
        newZoom = Math.min(Math.max(newZoom, 0.005f), 10f);
        return newZoom;
    }
}

File: src/client/kotlin/net/greenjab/fixedminecraft/map_book/MapBookScreen.kt
package net.greenjab.fixedminecraft.map_book

import net.greenjab.fixedminecraft.mixin.client.map.DrawContextAccessor
import net.greenjab.fixedminecraft.network.MapBookPlayer
import net.greenjab.fixedminecraft.registry.registries.ItemRegistry
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookItem
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookState
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookStateManager
import net.greenjab.fixedminecraft.registry.item.map_book.MapStateData
import net.minecraft.client.MinecraftClient
import net.minecraft.client.font.TextRenderer.TextLayerType
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.widget.ButtonWidget
import net.minecraft.client.render.LightmapTextureManager
import net.minecraft.client.render.RenderLayer
import net.minecraft.component.DataComponentTypes
import net.minecraft.component.type.MapIdComponent
import net.minecraft.item.ItemStack
import net.minecraft.item.map.MapDecoration
import net.minecraft.item.map.MapDecorationTypes
import net.minecraft.screen.ScreenTexts
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.RotationAxis
import net.minecraft.util.math.Vec3d
import net.minecraft.world.World
import java.util.Objects
import java.util.Optional
import kotlin.math.abs


class MapBookScreen(var item: ItemStack) : Screen(item.name) {
    var x = 0.0
    var y = 0.0
    var scale = 1.0f
    private var targetScale = 0.5f

    override fun init() {
        scale = 1.0f
        targetScale = 0.5f
        if (client != null && client!!.player != null) {
            x = -client!!.player!!.x
            y = -client!!.player!!.z
        }
        setScale(targetScale, width/2.0, height/2.0)
        for (mapStateData in (ItemRegistry.MAP_BOOK as MapBookItem).getMapStates(item, client?.world as World)) {
            addDrawable(MapTile(this, mapStateData.id, mapStateData.mapState, client!!))
        }
        addDrawableChild(ButtonWidget.builder(ScreenTexts.DONE) { this.close() }.dimensions(width / 2 - 100, height -40, 200, 20).build())
    }

    override fun shouldPause(): Boolean {
        return false
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        if (button == 2) {
            if (client?.player?.abilities?.creativeMode == true) {
                var pos = Vec3d(mouseX, mouseY, 0.0)
                pos = pos.multiply((1/scale).toDouble())
                pos = pos.subtract(width / 2.0, height / 2.0, 0.0)
                pos = pos.subtract(this.x/scale, this.y/scale, 0.0)
                MinecraftClient.getInstance().networkHandler?.sendCommand(String.format(
                    "tp %.6f %.6f %.6f",
                    pos.getX(), client?.player?.y, pos.getY()))
            }
        }
        return super.mouseClicked(mouseX, mouseY, button)
    }

    override fun mouseDragged(mouseX: Double, mouseY: Double, button: Int, deltaX: Double, deltaY: Double): Boolean {
        x += deltaX
        y += deltaY
        return super.mouseDragged(mouseX, mouseY, button, deltaX, deltaY)
    }

    override fun mouseScrolled(mouseX: Double, mouseY: Double, horizontalAmount: Double, verticalAmount: Double): Boolean {
        if (verticalAmount != 0.0) {
            targetScale = zoom(scale, -verticalAmount.toFloat())
        }
        return super.mouseScrolled(mouseX, mouseY, horizontalAmount, verticalAmount)
    }

    override fun render(context: DrawContext?, mouseX: Int, mouseY: Int, delta: Float) {
        if (context == null) return

        if (scale != targetScale) {
            val newScale = MathHelper.lerp(delta, scale, targetScale)
            setScale(newScale, mouseX.toDouble(), mouseY.toDouble())
        }

        super.render(context, mouseX, mouseY, delta)

        val thisPlayer = client?.player ?: return

        var stack = client?.player?.mainHandStack
        if (stack != null) {
            if (stack.item !is MapBookItem) stack = client?.player?.offHandStack
        }
        if (stack == null) return

        var id = -1
        if (stack.contains(DataComponentTypes.MAP_ID)) {
            id = stack.get(DataComponentTypes.MAP_ID)?.id ?: -1
        }

        if (id != -1) {
            val p = MapBookPlayer()
            p.setPlayer(thisPlayer)
            val m = MapBookStateManager.INSTANCE.getClientMapBookState(id)?.players
            if (m != null) {
                try {
                    for (player in m) {
                        if (player.dimension == p.dimension) {
                            if (player.name != p.name) {
                                renderPlayerIcon(context, player)
                            }
                        }
                    }
                } catch (_: ConcurrentModificationException) {
                }
            }
            renderPlayerIcon(context, p)
            renderIcons(context)
            renderPosition(context, mouseX, mouseY)
        }
    }

    private fun renderPosition(context: DrawContext, mouseX: Int, mouseY: Int) {

        var pos = Vec3d(mouseX.toDouble(), mouseY.toDouble(), 0.0)
        pos = pos.multiply((1/scale).toDouble())
        pos = pos.subtract(width / 2.0, height / 2.0, 0.0)
        pos = pos.subtract(this.x/scale, this.y/scale, 0.0)


        val textRenderer = MinecraftClient.getInstance().textRenderer
        val text = ""+pos.getX().toInt() + ", " + pos.getY().toInt()
        val o = textRenderer.getWidth(text).toFloat()
        Objects.requireNonNull(textRenderer)
        context.matrices.push()

        context.matrices.translate(width / 2.0, height -60.0, 20.0)

        context.matrices.translate(-o / 2f, 8.0f, 0.1f)

        textRenderer.draw(
            text,
            0.0f,
            0.0f,
            -1,
            false,
            context.matrices.peek().positionMatrix,
            (context as DrawContextAccessor).vertexConsumers,
            TextLayerType.NORMAL,
            Int.MIN_VALUE,
            LightmapTextureManager.MAX_LIGHT_COORDINATE
        )
        context.matrices.pop()

    }

    private fun renderPlayerIcon(context: DrawContext, player: MapBookPlayer) {
        val x = player.x.toFloat()
        val z = player.z.toFloat()
        val rotation = player.yaw

        context.matrices.push()
        context.matrices.translate(this.x, this.y, 0.0)
        context.matrices.scale(this.scale, this.scale, 1.0f)
        context.matrices.translate(x + width/ 2.0, z + height / 2.0, 0.0)
        context.matrices.multiply(RotationAxis.POSITIVE_Z.rotationDegrees(rotation))
        context.matrices.scale(8.0f, 8.0f, -3.0f)
        context.matrices.translate(0f, 0f, -10.0f)
        context.matrices.scale(1f / this.scale, 1f / this.scale, 1.0f)
        val sprite = client!!.mapDecorationsAtlasManager.getSprite(
            MapDecoration(
                MapDecorationTypes.PLAYER,
                0.toByte(),
                0.toByte(),
                0.toByte(),
                Optional.empty()
            )
        )
        val g = sprite.minU
        val h = sprite.minV
        val l = sprite.maxU
        val m = sprite.maxV
        val matrix4f2 = context.matrices.peek().positionMatrix
        val vertexConsumer2 = (context as DrawContextAccessor).vertexConsumers.getBuffer(RenderLayer.getText(sprite.atlasId))
        vertexConsumer2.vertex(matrix4f2, -1.0f, 1.0f, -0.1f).color(255, 255, 255, 255).texture(g, h).light(15728880)
        vertexConsumer2.vertex(matrix4f2, 1.0f, 1.0f, -0.1f).color(255, 255, 255, 255).texture(l, h).light(15728880)
        vertexConsumer2.vertex(matrix4f2, 1.0f, -1.0f, -0.1f).color(255, 255, 255, 255).texture(l, m).light(15728880)
        vertexConsumer2.vertex(matrix4f2, -1.0f, -1.0f, -0.1f).color(255, 255, 255, 255).texture(g, m).light(15728880)
        context.matrices.pop()

        val textRenderer = MinecraftClient.getInstance().textRenderer
        val text = player.name
        val o = textRenderer.getWidth(text).toFloat()
        Objects.requireNonNull(textRenderer)
        context.matrices.push()

        context.matrices.translate(this.x, this.y, 15.0)
        context.matrices.scale(this.scale, this.scale, 1.0f)

        context.matrices.translate(x + width / 2.0, z + height / 2.0, 0.0)

        context.matrices.scale(1 / this.scale, 1 / this.scale, 1.0f)
        context.matrices.translate(-o / 2f, 8.0f, 0.1f)

        textRenderer.draw(
            text,
            0.0f,
            0.0f,
            -1,
            false,
            context.matrices.peek().positionMatrix,
            context.vertexConsumers,
            TextLayerType.NORMAL,
            Int.MIN_VALUE,
            LightmapTextureManager.MAX_LIGHT_COORDINATE
        )
        context.matrices.pop()
    }

    private fun renderIcons(context: DrawContext) {
        var k = 0

        val light = LightmapTextureManager.MAX_LIGHT_COORDINATE

        var stack = client?.player?.mainHandStack
        if (stack != null) {
            if (stack.item !is MapBookItem) stack = client?.player?.offHandStack
        }

        for (mapStateData in getMapStates(stack!!, client?.world as World)) {
            var render = 0.0
            if (client?.world!!.dimensionEntry.idAsString.toString() == mapStateData.mapState.dimension.value.toString())
                render = 1.0
            if (client?.world!!.dimensionEntry.idAsString.toString().contains("the_nether") && mapStateData.mapState.dimension.value.toString().contains("overworld"))
                render = 1/8.0
            if (render>0) {
            val var11: Iterator<*> = mapStateData.mapState.decorations.iterator()
                while (var11.hasNext()) {
                    val mapIcon: MapDecoration = var11.next() as MapDecoration
                    if (mapIcon.type.idAsString != "minecraft:player" && mapIcon.type.idAsString != "minecraft:player_off_map" && mapIcon.type.idAsString != "minecraft:player_off_limits") {
                        context.matrices.push()
                        context.matrices.translate(this.x, this.y, 0.0)
                        context.matrices.scale(this.scale, this.scale, 1.0f)
                        val mapScale = (1 shl mapStateData.mapState.scale.toInt()).toFloat()
                        val offset = 64f * mapScale
                        val x = (mapStateData.mapState.centerX.toDouble() - offset.toDouble() + (mapIcon.x + 128 + 1) * mapScale / 2) * render
                        val z = (mapStateData.mapState.centerZ.toDouble() - offset.toDouble() + (mapIcon.z + 128 + 1) * mapScale / 2) * render
                        context.matrices.translate(x + width.toDouble() / 2.0, z + height.toDouble() / 2.0, 0.0)
                        context.matrices.multiply(RotationAxis.POSITIVE_Z.rotationDegrees(180F))
                        context.matrices.scale(8.0f, 8.0f, -3.0f)
                        context.matrices.translate(0f, 0f, -10.0f)
                        context.matrices.scale(1f / this.scale, 1f / this.scale, 1.0f)
                        val sprite = client!!.mapDecorationsAtlasManager.getSprite(
                            MapDecoration(
                                mapIcon.type,
                                0.toByte(),
                                0.toByte(),
                                0.toByte(),
                                Optional.empty()
                            )
                        )
                        val g = sprite.minU
                        val h = sprite.minV
                        val l = sprite.maxU
                        val m = sprite.maxV
                        val matrix4f2 = context.matrices.peek().positionMatrix
                        val vertexConsumer2 =
                            (context as DrawContextAccessor).vertexConsumers.getBuffer(RenderLayer.getText(sprite.atlasId))
                        vertexConsumer2.vertex(matrix4f2, -1.0f, 1.0f, -0.1f).color(255, 255, 255, 255).texture(g, h).light(15728880)
                        vertexConsumer2.vertex(matrix4f2, 1.0f, 1.0f, -0.1f).color(255, 255, 255, 255).texture(l, h).light(15728880)
                        vertexConsumer2.vertex(matrix4f2, 1.0f, -1.0f, -0.1f).color(255, 255, 255, 255).texture(l, m).light(15728880)
                        vertexConsumer2.vertex(matrix4f2, -1.0f, -1.0f, -0.1f).color(255, 255, 255, 255).texture(g, m).light(15728880)
                        context.matrices.pop()


                        if (mapIcon.name.isPresent) {
                            val textRenderer = MinecraftClient.getInstance().textRenderer
                            val text = mapIcon.name.get()
                            val o = textRenderer.getWidth(text).toFloat()
                            Objects.requireNonNull(textRenderer)
                            context.matrices.push()
                            context.matrices.translate(this.x, this.y, 11.0)
                            context.matrices.scale(this.scale, this.scale, 1.0f)
                            val mapx = (mapStateData.mapState.centerX.toDouble() - offset.toDouble() + (mapIcon.x + 128 + 1) * mapScale / 2) * render
                            val mapz = (mapStateData.mapState.centerZ.toDouble() - offset.toDouble() + (mapIcon.z + 128 + 1) * mapScale / 2) * render
                            context.matrices.translate(mapx + width.toDouble() / 2.0, mapz + height.toDouble() / 2.0, 0.0)
                            context.matrices.scale(1 / this.scale, 1 / this.scale, 1.0f)
                            context.matrices.translate(-o / 2f, 8.0f, 0.1f)

                            textRenderer.draw(
                                text,
                                0.0f,
                                0.0f,
                                -1,
                                false,
                                context.matrices.peek().positionMatrix,
                                context.vertexConsumers,
                                TextLayerType.NORMAL,
                                Int.MIN_VALUE,
                                light
                            )
                            context.matrices.pop()
                        }
                        ++k
                    }
                }
            }
        }
    }


    private fun getMapStates(stack: ItemStack, world: World): ArrayList<MapStateData> {
        val list = ArrayList<MapStateData>()
        val mapBookState = getMapBookState(stack, world)

        if (mapBookState != null) {
            for (i in mapBookState.mapIDs) {
                val mapState = world.getMapState(MapIdComponent(i))
                if (mapState != null) {
                    list.add(MapStateData(MapIdComponent(i), mapState))
                }
            }
        }
        return list
    }

    private fun getMapBookState(stack: ItemStack, world: World): MapBookState? {
        val id = getMapBookId(stack) ?: return null

        if (world.isClient) {
            return MapBookStateManager.INSTANCE.getClientMapBookState(id)
        } else if (world.server != null) {
            return MapBookStateManager.INSTANCE.getMapBookState(world.server!!, id)
        }
        return null
    }

    private fun getMapBookId(stack: ItemStack): Int? {
        val mapIdComponent = stack.getOrDefault(DataComponentTypes.MAP_ID, null)
        return mapIdComponent?.id()
    }

    private fun setScale(newScale: Float, mouseX: Double, mouseY: Double) {
        val offsetX = x-mouseX
        val offsetY = y-mouseY

        val scaleChange = newScale/scale

        x = (scaleChange * offsetX)+mouseX
        y = (scaleChange * offsetY)+mouseY

        scale = newScale
    }

    private fun zoom(start: Float, scroll: Float): Float {
        // logarithmic zoom that doesn't drift when zooming in and out repeatedly
        val absScroll = abs(scroll)
        val speed = 5.0f
        var newZoom = if (scroll > 0) start - (start / (scroll * speed)) else (start * absScroll * speed) / (absScroll * speed - 1)
        newZoom = Math.clamp(newZoom, 0.005f, 10f)
        return newZoom
    }
}
